File: backend/schemas/expense.py
from pydantic import BaseModel, Field
from typing import Optional
from datetime import date, datetime
from uuid import UUID


class ExpenseCreate(BaseModel):
    """Schema for creating a new expense."""
    name: str = Field(..., min_length=1, max_length=255, description="Expense name")
    category: str = Field(..., min_length=1, max_length=100, description="Expense category")
    amount: float = Field(..., gt=0, description="Expense amount in RM")
    frequency: str = Field("one-time", description="Frequency: one-time, daily, weekly, monthly, yearly")
    transaction_date: Optional[date] = Field(None, description="Date of the expense")
    notes: Optional[str] = Field(None, max_length=500, description="Additional notes")
    description: Optional[str] = Field(None, max_length=500, description="Description")


class ExpenseUpdate(BaseModel):
    """Schema for updating an existing expense. All fields optional."""
    name: Optional[str] = Field(None, min_length=1, max_length=255)
    category: Optional[str] = Field(None, min_length=1, max_length=100)
    amount: Optional[float] = Field(None, gt=0)
    frequency: Optional[str] = Field(None)
    transaction_date: Optional[date] = Field(None)
    notes: Optional[str] = Field(None, max_length=500)
    description: Optional[str] = Field(None, max_length=500)


class ExpenseResponse(BaseModel):
    """Schema for expense responses from the API."""
    id: UUID
    business_id: Optional[UUID] = None
    name: Optional[str] = None
    category: Optional[str] = None
    amount: float
    frequency: Optional[str] = "one-time"
    transaction_date: Optional[date] = None
    description: Optional[str] = None
    notes: Optional[str] = None
    created_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class BudgetCreate(BaseModel):
    """Schema for creating/updating a budget allocation."""
    category: str = Field(..., min_length=1, max_length=100, description="Budget category")
    allocated_amount: float = Field(..., ge=0, description="Allocated budget in RM")
    month: int = Field(..., ge=1, le=12, description="Month (1-12)")
    year: int = Field(..., ge=2020, description="Year")


class BudgetUpdate(BaseModel):
    """Schema for updating a budget allocation."""
    allocated_amount: Optional[float] = Field(None, ge=0)
    month: Optional[int] = Field(None, ge=1, le=12)
    year: Optional[int] = Field(None, ge=2020)


class BudgetResponse(BaseModel):
    """Schema for budget responses from the API."""
    id: UUID
    business_id: Optional[UUID] = None
    category: str
    allocated_amount: float
    month: int
    year: int
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True

File: frontend/src/types/expense.ts
// ─── Expense Types ─────────────────────────────────────────────────────

export interface Expense {
  id: string;
  business_id: string | null;
  name: string | null;
  category: string | null;
  amount: number;
  frequency: string | null;
  transaction_date: string | null;
  description: string | null;
  notes: string | null;
  created_at: string | null;
}

export interface ExpenseFormData {
  name: string;
  category: string;
  amount: number;
  frequency: string;
  transaction_date: string;
  notes: string;
}

// ─── Budget Types ──────────────────────────────────────────────────────

export interface Budget {
  id: string;
  business_id: string | null;
  category: string;
  allocated_amount: number;
  month: number;
  year: number;
  created_at: string | null;
  updated_at: string | null;
}

export interface BudgetFormData {
  category: string;
  allocated_amount: number;
  month: number;
  year: number;
}

// ─── Summary Types ─────────────────────────────────────────────────────

export interface MonthlySummary {
  total_spent: number;
  by_category: Record<string, number>;
  biggest_category: string | null;
  expense_count: number;
}

export interface MonthComparison {
  month: number;
  year: number;
  label: string;
  total_spent: number;
  by_category: Record<string, number>;
}

// ─── Budget Status ─────────────────────────────────────────────────────

export type BudgetStatus = 'over' | 'near' | 'on-track';

export function getBudgetStatus(spent: number, allocated: number): BudgetStatus {
  if (allocated <= 0) return 'on-track';
  const pct = (spent / allocated) * 100;
  if (pct > 100) return 'over';
  if (pct > 80) return 'near';
  return 'on-track';
}

export function getBudgetPercentage(spent: number, allocated: number): number {
  if (allocated <= 0) return 0;
  return Math.round((spent / allocated) * 100);
}

// ─── Constants ─────────────────────────────────────────────────────────

export const EXPENSE_CATEGORIES = [
  { value: 'ingredients', label: 'Ingredients' },
  { value: 'utilities', label: 'Utilities' },
  { value: 'rent', label: 'Rent' },
  { value: 'salaries', label: 'Salaries' },
  { value: 'marketing', label: 'Marketing' },
  { value: 'equipment', label: 'Equipment' },
  { value: 'subscriptions', label: 'Subscriptions' },
  { value: 'maintenance', label: 'Maintenance' },
  { value: 'transport', label: 'Transport' },
  { value: 'other', label: 'Other' },
] as const;

export const FREQUENCY_OPTIONS = [
  { value: 'one-time', label: 'One-time' },
  { value: 'daily', label: 'Daily' },
  { value: 'weekly', label: 'Weekly' },
  { value: 'monthly', label: 'Monthly' },
  { value: 'yearly', label: 'Yearly' },
] as const;

export const MONTH_NAMES = [
  'January', 'February', 'March', 'April', 'May', 'June',
  'July', 'August', 'September', 'October', 'November', 'December',
] as const;

export type ExpenseCategory = (typeof EXPENSE_CATEGORIES)[number]['value'];
export type FrequencyType = (typeof FREQUENCY_OPTIONS)[number]['value'];
